### DISCUSSION TITLE
Using Div.(Grad v) term in derivativeparsedmaterial

### DISCUSSION THREAD

**Discussion by vermaprk** [idaholab/moose#28385] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello
I need to insert the diffusion term value `(Div.Grad v)/|Grad v|` in the derivativeparsedmaterial. Here is my approach:

I crearted a kernel that calculates (Div.Grad v)/|Grad v| where v is the coupled variable and the kernel variable is defined as curr.
Now I am using curr as a coupled_variable in the DerivativeParsedmaterial to evaluate the material.

Couple of questions:
Is this the right approach?  I tried using only material to calculate the `(Div.Grad v)/|Grad v|`. However, I could not find the way to resolve the divergence part as there is no test function.
I also tried using auxvariable but was not able to do this.

Thanks

**Comment by GiudGiud** [comment]
Hello

> I crearted a kernel that calculates (Div.Grad v)/|Grad v| where v is the coupled variable and the kernel variable is defined as curr.

do you mean an auxiliary kernel (auxkernel)? Kernels compute contributions to the residual, only auxkernels can set variable values

> However, I could not find the way to resolve the divergence part as there is no test function.

why do you need the test function?

**Reply by vermaprk** [reply]
> why do you need the test function?

Div.Grad v is there that means double differentiation. So, I was thinking if I can write this as auxkernel  as
`return - (_grad_test[_i][_qp] * _grad_v[_qp])/_grad_v[_qp].norm();`
However, as `computeQpResidual()` does not work for auxkernel like the kernel system. I was wondering if there was some way out.

Otherwise how to get double differentiation in auxkernel. 
`_gradient(coupledGradient("gradient_variable")) `only differentiates once. Please provide the syntax to get a double differential
Thanks

**Reply by GiudGiud** [reply]
coupledSecond gets you the double derivatives
in the Coupleable interface (coupleable.h), same as coupledGradient
https://mooseframework.inl.gov/source/interfaces/Coupleable.html

**Reply by vermaprk** [reply]
OK as per your suggestion I tried:
```
Real
EgCurvatureValue::computeValue()
{
  return _grad2[_qp](0,0);
}
```
So this should return d2v/dx2 as I assume (0,0) means differentiating w.r.t x and x. But I don't get the required solution. I have tested this for steady-state diffusion problems with source term 1. I should get 1 as an answer, but I am getting 0.
My .C file is as follow:
```
#include "EgCurvatureValue.h"

registerMooseObject("goldfishApp", EgCurvatureValue);

InputParameters
EgCurvatureValue::validParams()
{
  InputParameters params = AuxKernel::validParams();
  params.addClassDescription(
      "Creates a field consisting of one component of the gradient of a coupled variable.");
  params.addRequiredCoupledVar("gradient_variable",
                               "The variable from which to compute the gradient component");
  return params;
}

EgCurvatureValue::EgCurvatureValue(const InputParameters & parameters)
  : AuxKernel(parameters),
    _grad(coupledGradient("gradient_variable")),
    _grad2(coupledSecond("gradient_variable"))
{
}

Real
EgCurvatureValue::computeValue()
{
  return _grad2[_qp](0,0);
}
```
Thanks

**Reply by GiudGiud** [reply]
what finite element fammily are you using?
anything first order has a second derivative of 0
you'll need to go higher order, (or code a multi-element stencil, difficult) to get non-zero second derivatives

**Comment by vermaprk** [comment]
I am using second order monomial variable.
I have tried going higher order but still getting zero.

INPUT FILE
```
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 100
  xmax =100
[]

[Variables]
  [./diffused]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[AuxVariables]
  [./prad]
    order = SECOND
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = diffused
  [../]
   [./body_force]
    type = BodyForce
    variable = diffused
    value = -1
  [../]
[]
[AuxKernels]
  [./prad]
    type = EgCurvatureValue
    variable = prad
    gradient_variable = diffused
  [../]
[]


[BCs]
  [./left]
    type = DirichletBC
    variable = diffused
    boundary = 'left'
    value = 0
  [../]

  [./right]
    type = DirichletBC
    variable = diffused
    boundary = 'right'
    value = 0
  [../]
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'
[]

[Outputs]
  execute_on = 'FINAL'
  exodus = true
[]
```

**Reply by GiudGiud** [reply]
can you paste your header here?
I want to make sure you are capturing these as references

**Reply by vermaprk** [reply]
```
Framework Information:
MOOSE Version:           git commit 30fe3abb9c on 2024-07-31
LibMesh Version:         
PETSc Version:           3.20.3
SLEPc Version:           3.20.1
Current Time:            Fri Aug 16 22:51:20 2024
Executable Timestamp:    Fri Aug 16 22:34:47 2024

Checkpoint:
  Wall Time Interval:      Every 3600.000000 s
  User Checkpoint:         Disabled
  # Checkpoints Kept:      2
  Execute On:              TIMESTEP_END 

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          1
  Spatial Dimension:       1
  Nodes:                   101
  Elems:                   100
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                101
  Num Local DOFs:          101
  Variables:               "diffused" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                300
  Num Local DOFs:          300
  Variables:               "prad" 
  Finite Element Types:    "MONOMIAL" 
  Approximation Orders:    "SECOND" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             Preconditioned JFNK
```

**Reply by GiudGiud** [reply]
the header as in the .h file please. Though I understand this is the simulation header.

```
  Variables:               "diffused" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 
```
thre source variable is first order.  This one will need to be second order too

**Reply by vermaprk** [reply]
here is the header file
```
#pragma once

// MOOSE includes
#include "AuxKernel.h"

/**
 * Extract a component from the gradient of a variable
 */
class EgCurvatureValue : public AuxKernel
{
public:
  static InputParameters validParams();

  /**
   * Class constructor
   * @param parameters Input parameters for the object
   */
  EgCurvatureValue(const InputParameters & parameters);

protected:
  virtual Real computeValue() override;

private:
  /// Reference to the gradient of the coupled variable
  const VariableGradient & _grad;
  const VariableSecond & _grad2;
};
```
second order lagrange has issues:
```
*** ERROR ***
ERROR: Bad ElemType = EDGE2 for SECOND order approximation!
Stack frames: 18
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
2: /home/prk/miniforge/envs/moose/libmesh/lib/libmesh_opt.so.0(+0x8902d9) [0x7fd05c0912d9]
3: libMesh::DofMap::reinit(libMesh::MeshBase&)
4: libMesh::DofMap::distribute_dofs(libMesh::MeshBase&)
5: libMesh::System::init_data()
6: libMesh::System::init()
7: libMesh::EquationSystems::init()
8: FEProblemBase::init()
9: Action::timedAct()
10: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
11: ActionWarehouse::executeAllActions()
12: MooseApp::runInputFile()
13: MooseApp::run()
14: main
15: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fd050a49d90]
16: __libc_start_main
17: ./goldfish-opt(+0x33ef) [0x7fd061fad3ef]
[0] ../src/fe/fe_lagrange.C, line 789, compiled Jul 30 2024 at 17:35:57
```

**Reply by GiudGiud** [reply, marked as answer]
in the [Mesh] block, add second_order = true
